### ISSUE TITLE
Deleting Enclosing parent causes 500 error on children

### ISSUE THREAD

**Issue by andrewvaughan** [gramps-project/gramps-web-api#471]
If you delete a Place that is registered as the "Enclosing" parent, it causes all references to any children to throw a 500 error, including maps, etc:

```
gramps-app | [2023-12-26 21:22:32 +0000] [24] [ERROR] Error handling request /api/places/?gramps_id=P0000&backlinks=true&extend=all&locale=en&profile=all
gramps-app | Traceback (most recent call last):
gramps-app |   File "/usr/local/lib/python3.11/dist-packages/gunicorn/workers/sync.py", line 135, in handle
gramps-app |     self.handle_request(listener, req, client, addr)
gramps-app |   File "/usr/local/lib/python3.11/dist-packages/gunicorn/workers/sync.py", line 178, in handle_request
gramps-app |     respiter = self.wsgi(environ, resp.start_response)
gramps-app |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
gramps-app |   File "/usr/local/lib/python3.11/dist-packages/flask/app.py", line 1478, in __call__
gramps-app |     return self.wsgi_app(environ, start_response)
gramps-app |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
gramps-app |   File "/usr/local/lib/python3.11/dist-packages/flask/app.py", line 1458, in wsgi_app
gramps-app |     response = self.handle_exception(e)
gramps-app |                ^^^^^^^^^^^^^^^^^^^^^^^^
gramps-app |   File "/usr/local/lib/python3.11/dist-packages/flask/app.py", line 1455, in wsgi_app
gramps-app |     response = self.full_dispatch_request()
gramps-app |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
gramps-app |   File "/usr/local/lib/python3.11/dist-packages/flask/app.py", line 869, in full_dispatch_request
gramps-app |     rv = self.handle_user_exception(e)
gramps-app |          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
gramps-app |   File "/usr/local/lib/python3.11/dist-packages/flask/app.py", line 867, in full_dispatch_request
gramps-app |     rv = self.dispatch_request()
gramps-app |          ^^^^^^^^^^^^^^^^^^^^^^^
gramps-app |   File "/usr/local/lib/python3.11/dist-packages/flask/app.py", line 852, in dispatch_request
gramps-app |     return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)
gramps-app |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
gramps-app |   File "/usr/local/lib/python3.11/dist-packages/gramps_webapi/api/auth.py", line 44, in wrapper
gramps-app |     return func(*args, **kwargs)
gramps-app |            ^^^^^^^^^^^^^^^^^^^^^
gramps-app |   File "/usr/local/lib/python3.11/dist-packages/flask/views.py", line 109, in view
gramps-app |     return current_app.ensure_sync(self.dispatch_request)(**kwargs)
gramps-app |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
gramps-app |   File "/usr/local/lib/python3.11/dist-packages/flask/views.py", line 190, in dispatch_request
gramps-app |     return current_app.ensure_sync(meth)(**kwargs)
gramps-app |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
gramps-app |   File "/usr/local/lib/python3.11/dist-packages/webargs/core.py", line 649, in wrapper
gramps-app |     return func(*args, **kwargs)
gramps-app |            ^^^^^^^^^^^^^^^^^^^^^
gramps-app |   File "/usr/local/lib/python3.11/dist-packages/gramps_webapi/api/resources/base.py", line 387, in get
gramps-app |     200, [self.full_object(obj, args, locale=locale)], args, total_items=1
gramps-app |           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
gramps-app |   File "/usr/local/lib/python3.11/dist-packages/gramps_webapi/api/resources/base.py", line 91, in full_object
gramps-app |     obj = self.object_extend(obj, args, locale=locale)
gramps-app |           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
gramps-app |   File "/usr/local/lib/python3.11/dist-packages/gramps_webapi/api/resources/places.py", line 48, in object_extend
gramps-app |     obj.profile = get_place_profile_for_object(
gramps-app |                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
gramps-app |   File "/usr/local/lib/python3.11/dist-packages/gramps_webapi/api/resources/util.py", line 361, in get_place_profile_for_object
gramps-app |     _place = db_handle.get_place_from_handle(handle)
gramps-app |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
gramps-app |   File "/usr/lib/python3/dist-packages/gramps/gen/db/generic.py", line 1285, in get_place_from_handle
gramps-app |     return self._get_from_handle(PLACE_KEY, Place, handle)
gramps-app |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
gramps-app |   File "/usr/lib/python3/dist-packages/gramps/gen/db/generic.py", line 1270, in _get_from_handle
gramps-app |     raise HandleError('Handle %s not found' % handle)
gramps-app | gramps.gen.errors.HandleError: Handle f7e8f271b09468cc754718a7255 not found
```

**Comment by DavidMStraub** [comment]
I think there are two problems here:

- we should be more robust in handling (i.e. ignoring) non-existing handles
- the `delete_place` function does not clean up place references

By the way, Gramps Desktop treats this differently: it doesn't even allow deleting parent places! When you try, it throws

> This place is currently referenced by another place. First remove the places it contains.

This is understandable as it might be dangerous deleting a place (e.g. a country) that is the parent of hundreds of other places accidentally.

Opinions on how this should be handled ideally?

**Comment by PQYPLZXHGF** [comment]
I would expect Gramps Web to do the same thing Gramps Desktop does in this instance and generally.

Thanks for Gramps web by the web 🥇

**Comment by DavidMStraub** [comment]
Yes, that makes sense. Even so, I still wonder whether we should make this a pure UI restriction (i.e., graying out the delete button) or forbid deletion in the API itself. Right now, deletion is allowed, but place references are not cleaned up. It would be fairly easy to add cleanup of place references and it also seems consistent as we also remove place references in all other objects, like events, upon deletion.

**Comment by andrewvaughan** [comment]
IMO ideally there should be a foreign key constraint in the database that bubbles up as an error. That way it doesn't matter who or what is making the attempt.

**Comment by DavidMStraub** [comment]
That's unfortunately not possible because of the way Gramps uses the database. For historical reasons, the data is not stored properly relationally, so relationships between objects are not foreign keys but just metadata stored in a blob.

**Comment by emyoulation** [comment]
Instead of Deleting the Place that encloses other Places, could you merge the Place Upwards into its enclosing Place? 

That should have a logical resolution for all but the Level1 (or is it zero?) place

**Comment by DavidMStraub** [comment]
This error 500 should be fixed in 1.5.1. There is still a question whether we should allow deletion of parent places, but I think that's a frontend question.
